File: languages/python/03-advanced/copy-and-move-semantics/example/main.py
# Module focus: How copying, sharing, or transferring state changes later behavior.
# Why it matters: the example makes it possible to predict aliasing and independence after
# copying or sharing values before the learner tackles the exercises.

# Separate helpers keep the main path focused on how to predict aliasing and independence after
# copying or sharing values.
class Buffer:
    def __init__(self, size: int) -> None:
        # These values exercise the normal path before the exercises vary the documented
        # boundaries.
        safe_size = max(0, size)
        self._values = [0] * safe_size
        # The printed result shows whether the program can choose an idiomatic ownership-transfer
        # strategy for the language.
        print(f"Constructed (size={len(self._values)})")

    def clone(self) -> "Buffer":
        cloned = Buffer.__new__(Buffer)
        cloned._values = self._values.copy()
        print("Cloned")
        return cloned

    def transfer(self) -> "Buffer":
        moved_values = self._values
        self._values = []
        transferred = Buffer.__new__(Buffer)
        transferred._values = moved_values
        print(f"Transferred (size={len(moved_values)})")
        return transferred

    @property
    def size(self) -> int:
        return len(self._values)


# Fixed inputs make the consequence of assuming `=` duplicates list contents visible and
# repeatable.
def main() -> None:
    first = Buffer(3)
    second = first.clone()
    third = second.transfer()

    print(f"first size: {first.size}")
    print(f"second size (after transfer): {second.size}")
    print(f"third size: {third.size}")


if __name__ == "__main__":
    main()

File: languages/csharp/03-advanced/copy-and-move-semantics/example/main.cs
// Module focus: How copying, sharing, or transferring state changes later behavior.
// Why it matters: the example makes it possible to predict aliasing and independence after
// copying or sharing values before the learner tackles the exercises.

using System;
using System.Collections.Generic;

// Separate helpers keep the main path focused on how to predict aliasing and independence after
// copying or sharing values.
class Buffer
{
    private List<int> values;

    public Buffer(int size)
    {
        int safeSize = size < 0 ? 0 : size;
        values = new List<int>(new int[safeSize]);
        Console.WriteLine($"Constructed (size={values.Count})");
    }

    private Buffer(List<int> sourceValues)
    {
        values = sourceValues;
    }

    public Buffer Clone()
    {
        Buffer copy = new Buffer(new List<int>(values));
        Console.WriteLine("Cloned");
        return copy;
    }

    public Buffer Transfer()
    {
        List<int> movedValues = values;
        values = new List<int>();
        Buffer transferred = new Buffer(movedValues);
        Console.WriteLine($"Transferred (size={transferred.Size})");
        return transferred;
    }

    public int Size => values.Count;
}

class Program
{
    // Fixed inputs make the consequence of assuming assignment duplicates object contents for
    // classes visible and repeatable.
    static void Main()
    {
        // These values exercise the normal path before the exercises vary the documented
        // boundaries.
        Buffer first = new Buffer(3);
        Buffer second = first.Clone();
        Buffer third = second.Transfer();

        // The printed result shows whether the program can choose an idiomatic ownership-transfer
        // strategy for the language.
        Console.WriteLine($"first size: {first.Size}");
        Console.WriteLine($"second size (after transfer): {second.Size}");
        Console.WriteLine($"third size: {third.Size}");
    }
}
